File: android/src/main/java/com/nextimage/NextImageSecurity.kt
package com.nextimage

/**
 * URL and header hardening, mirroring `src/security.ts`.
 *
 * This file deliberately has no Android or React Native imports so it can be
 * exercised by plain JVM unit tests. The JS layer validates first; this is the
 * second line of defence for anything that reaches native through a preload,
 * a stale bundle, or a direct native caller.
 */
object NextImageSecurity {

  /** Headers that would let a caller rewrite the request line or break framing. */
  private val FORBIDDEN_HEADERS = setOf(
    "connection",
    "content-length",
    "expect",
    "host",
    "keep-alive",
    "proxy-authorization",
    "proxy-connection",
    "te",
    "trailer",
    "transfer-encoding",
    "upgrade",
    "via",
  )

  /**
   * Schemes Coil loads without a network connection. `require()`d assets do
   * not come through here: the JS layer marks them as bundled and the request
   * factory resolves them directly.
   */
  private val LOCAL_SCHEMES = setOf(
    "android.resource",
    "content",
  )

  private val HEADER_NAME_REGEX = Regex("^[!#$%&'*+\\-.^_`|~0-9A-Za-z]+$")
  private val IPV4_REGEX = Regex("^(\\d{1,3})\\.(\\d{1,3})\\.(\\d{1,3})\\.(\\d{1,3})$")
  private val URI_REGEX =
    Regex("^([a-zA-Z][a-zA-Z0-9+.\\-]*):(//([^/?#]*))?([^?#]*)(\\?([^#]*))?(#(.*))?$")

  data class ParsedUri(
    val scheme: String,
    val userInfo: String?,
    val host: String?,
    val port: Int?,
    val path: String,
    val query: String?,
  )

  sealed class UriResult {
    data class Allowed(val uri: String, val parsed: ParsedUri) : UriResult()
    data class Blocked(val code: String, val message: String) : UriResult()
  }

  fun parseUri(uri: String): ParsedUri? {
    val match = URI_REGEX.matchEntire(uri) ?: return null
    val scheme = match.groupValues[1].lowercase()
    val hasAuthority = match.groups[2] != null
    var userInfo: String? = null
    var host: String? = null
    var port: Int? = null

    if (hasAuthority) {
      var rest = match.groupValues[3]
      val at = rest.lastIndexOf('@')
      if (at != -1) {
        userInfo = rest.substring(0, at)
        rest = rest.substring(at + 1)
      }

      if (rest.startsWith("[")) {
        val close = rest.indexOf(']')
        if (close == -1) return null
        host = rest.substring(0, close + 1).lowercase()
        rest = rest.substring(close + 1)
        if (rest.isNotEmpty() && !rest.startsWith(":")) return null
      } else {
        val colon = rest.indexOf(':')
        host = (if (colon == -1) rest else rest.substring(0, colon)).lowercase()
        rest = if (colon == -1) "" else rest.substring(colon)
      }

      if (rest.startsWith(":")) {
        val digits = rest.substring(1)
        if (digits.isNotEmpty()) {
          val parsedPort = digits.toIntOrNull() ?: return null
          if (parsedPort < 1 || parsedPort > 65535) return null
          port = parsedPort
        }
      }

      if (host.isEmpty()) host = null
    }

    return ParsedUri(
      scheme = scheme,
      userInfo = userInfo,
      host = host,
      port = port,
      path = match.groupValues[4],
      query = match.groups[6]?.value,
    )
  }

  /** Supports `example.com`, `.example.com` and `*.example.com`. */
  fun hostMatches(host: String, pattern: String): Boolean {
    val normalizedHost = host.lowercase().trim('[', ']')
    var normalizedPattern = pattern.lowercase().trim()
    if (normalizedPattern.isEmpty()) return false
    if (normalizedPattern == "*") return true
    if (normalizedPattern.startsWith("*.")) {
      normalizedPattern = normalizedPattern.substring(1)
    }
    if (normalizedPattern.startsWith(".")) {
      val bare = normalizedPattern.substring(1)
      return normalizedHost == bare || normalizedHost.endsWith(normalizedPattern)
    }
    return normalizedHost == normalizedPattern
  }

  /** Loopback, link-local, private-range, CGNAT and cloud metadata hosts. */
  fun isPrivateHost(host: String): Boolean {
    val normalized = host.lowercase().trim('[', ']').trimEnd('.')

    if (normalized == "localhost" ||
      normalized.endsWith(".localhost") ||
      normalized.endsWith(".local") ||
      normalized.endsWith(".internal")
    ) {
      return true
    }

    val ipv4 = IPV4_REGEX.matchEntire(normalized)
    if (ipv4 != null) {
      val octets = (1..4).map { ipv4.groupValues[it].toInt() }
      if (octets.any { it > 255 }) return true
      val a = octets[0]
      val b = octets[1]
      return when {
        a == 0 || a == 10 || a == 127 -> true
        a == 169 && b == 254 -> true
        a == 172 && b in 16..31 -> true
        a == 192 && b == 168 -> true
        a == 100 && b in 64..127 -> true
        a >= 224 -> true
        else -> false
      }
    }

    if (normalized.contains(':')) {
      if (normalized == "::" || normalized == "::1") return true
      if (normalized.startsWith("::ffff:")) {
        return isPrivateHost(normalized.removePrefix("::ffff:"))
      }
      if (Regex("^f[cd][0-9a-f]{0,2}:").containsMatchIn(normalized)) return true
      if (Regex("^fe[89ab][0-9a-f]?:").containsMatchIn(normalized)) return true
    }

    return false
  }

  fun validateUri(rawUri: String?, config: NextImageConfig): UriResult {
    if (rawUri.isNullOrBlank()) {
      return UriResult.Blocked("EMPTY_URI", "Image source uri must not be empty.")
    }

    val uri = rawUri.trim()
    if (uri.length > config.maxUriLength) {
      return UriResult.Blocked(
        "URI_TOO_LONG",
        "Image source uri exceeds ${config.maxUriLength} characters.",
      )
    }
    if (uri.any { it.code < 0x20 || it.code == 0x7f }) {
      return UriResult.Blocked(
        "CONTROL_CHARACTERS",
        "Image source uri contains control characters.",
      )
    }

    val parsed = parseUri(uri)
      ?: return UriResult.Blocked("MALFORMED_URI", "Image source uri is malformed.")

    when (parsed.scheme) {
      "https" -> Unit
      "http" -> if (!config.allowInsecureHttp) {
        return UriResult.Blocked(
          "INSECURE_SCHEME",
          "Plain http:// sources are blocked by the NextImage security policy.",
        )
      }
      "data" -> return when {
        !config.allowDataUri ->
          UriResult.Blocked("DATA_URI_NOT_ALLOWED", "data: sources are blocked.")
        estimateDataUriBytes(uri) > config.maxDataUriBytes ->
          UriResult.Blocked(
            "DATA_URI_TOO_LARGE",
            "data: source exceeds ${config.maxDataUriBytes} bytes.",
          )
        else -> UriResult.Allowed(uri, parsed)
      }
      "file" -> return if (config.allowFileUri) {
        UriResult.Allowed(uri, parsed)
      } else {
        UriResult.Blocked("FILE_URI_NOT_ALLOWED", "file: sources are blocked.")
      }
      else -> return if (LOCAL_SCHEMES.contains(parsed.scheme)) {
        UriResult.Allowed(uri, parsed)
      } else {
        UriResult.Blocked(
          "SCHEME_NOT_ALLOWED",
          "Unsupported uri scheme \"${parsed.scheme}\".",
        )
      }
    }

    if (parsed.userInfo != null && !config.allowUriCredentials) {
      return UriResult.Blocked(
        "URI_CREDENTIALS",
        "Image source uri must not embed credentials.",
      )
    }

    val host = parsed.host
      ?: return UriResult.Blocked("HOST_MISSING", "Image source uri has no host.")

    if (config.blockedHosts.any { hostMatches(host, it) }) {
      return UriResult.Blocked("HOST_BLOCKED", "Host \"$host\" is blocked.")
    }
    val allowed = config.allowedHosts
    if (allowed != null && !allowed.any { hostMatches(host, it) }) {
      return UriResult.Blocked("HOST_NOT_ALLOWED", "Host \"$host\" is not allow-listed.")
    }
    if (config.blockPrivateNetworks && isPrivateHost(host)) {
      return UriResult.Blocked(
        "PRIVATE_HOST_BLOCKED",
        "Host \"$host\" resolves to a private or loopback address.",
      )
    }

    return UriResult.Allowed(uri, parsed)
  }

  data class SanitizedHeaders(
    val headers: Map<String, String>,
    val rejected: List<String>,
  )

  /**
   * Remove headers that could split the request or that the HTTP client owns.
   * Values keep their order; duplicates resolve to the last value seen.
   */
  fun sanitizeHeaders(
    headers: List<Pair<String, String>>,
    config: NextImageConfig,
  ): SanitizedHeaders {
    val accepted = LinkedHashMap<String, String>()
    val rejected = mutableListOf<String>()

    for ((name, value) in headers) {
      if (accepted.size >= config.maxHeaderCount) {
        rejected.add(name)
        continue
      }
      if (name.length > config.maxHeaderNameLength ||
        !HEADER_NAME_REGEX.matches(name) ||
        FORBIDDEN_HEADERS.contains(name.lowercase())
      ) {
        rejected.add(name)
        continue
      }
      if (value.length > config.maxHeaderValueLength || !isValidHeaderValue(value)) {
        rejected.add(name)
        continue
      }
      accepted[name] = value
    }

    return SanitizedHeaders(accepted, rejected)
  }

  /** Decoded length of a data uri, without decoding it. */
  private fun estimateDataUriBytes(uri: String): Long {
    val comma = uri.indexOf(',')
    if (comma == -1) return uri.length.toLong()
    val meta = uri.substring(0, comma)
    val payload = uri.substring(comma + 1)
    if (!meta.endsWith(";base64", ignoreCase = true)) return payload.length.toLong()
    val padding = when {
      payload.endsWith("==") -> 2
      payload.endsWith("=") -> 1
      else -> 0
    }
    return (payload.length * 3L / 4L - padding).coerceAtLeast(0L)
  }

  private fun isValidHeaderValue(value: String): Boolean = value.all { char ->
    val code = char.code
    char == '\t' || (code in 0x20..0x7e) || (code in 0xa0..0xff)
  }

  /** Credential-free, query-free rendering of a uri, safe for logs and events. */
  fun redactUri(uri: String): String {
    val parsed = parseUri(uri) ?: return "<invalid-uri>"
    if (parsed.scheme == "data") return "data:<redacted>"
    val port = parsed.port?.let { ":$it" } ?: ""
    val query = if (parsed.query == null) "" else "?<redacted>"
    return "${parsed.scheme}://${parsed.host ?: ""}$port${parsed.path}$query"
  }
}

File: ios/NextImageSecurity.swift
import Foundation

/// Runtime configuration, set from JS through `NextImage.configure(...)`.
///
/// Foundation only, with no UIKit or Kingfisher imports, so the same file is
/// compiled and asserted against by `tests/ios` on every CI run.
public struct NextImageConfig {
    public static let defaultDiskCacheBytes: UInt = 250 * 1024 * 1024
    public static let defaultMemoryCacheBytes: UInt = 0 // 0 means "let the OS decide"
    public static let defaultCacheDurationMinutes: Double = 10080
    public static let minDiskCacheBytes: UInt = 4 * 1024 * 1024
    public static let minMemoryCacheBytes: UInt = 1 * 1024 * 1024

    public var allowInsecureHttp = false
    public var allowDataUri = false
    public var allowFileUri = false
    /// `nil` means "no allow-list".
    public var allowedHosts: [String]?
    public var blockedHosts: [String] = []
    public var blockPrivateNetworks = true
    public var allowUriCredentials = false
    public var maxUriLength = 8192
    public var maxDataUriBytes = 2 * 1024 * 1024
    public var maxHeaderCount = 24
    public var maxHeaderNameLength = 128
    public var maxHeaderValueLength = 8192
    /// SPKI pins per host in `sha256/<base64>` form. Empty disables pinning.
    public var certificatePins: [String: [String]] = [:]
    public var memoryCacheBytes = NextImageConfig.defaultMemoryCacheBytes
    public var diskCacheBytes = NextImageConfig.defaultDiskCacheBytes
    public var requestTimeoutMs: Double = 30000
    /// When false, `cacheDuration` decides the lifetime, not the server.
    public var respectServerCacheHeaders = false

    public init() {}

    /// Changes to these fields require the downloader and cache to be rebuilt.
    public static func requiresLoaderRebuild(_ before: NextImageConfig, _ after: NextImageConfig) -> Bool {
        before.memoryCacheBytes != after.memoryCacheBytes
            || before.diskCacheBytes != after.diskCacheBytes
            || before.requestTimeoutMs != after.requestTimeoutMs
            || before.certificatePins != after.certificatePins
    }

    private static let pinPattern = "^sha256/[A-Za-z0-9+/]{43}=$"

    /// Applies a partial configuration coming from JS. Values of the wrong
    /// type, negative sizes and malformed pins are ignored rather than trusted.
    public mutating func apply(options: [String: Any]) {
        if let value = options["allowInsecureHttp"] as? Bool { allowInsecureHttp = value }
        if let value = options["allowDataUri"] as? Bool { allowDataUri = value }
        if let value = options["allowFileUri"] as? Bool { allowFileUri = value }
        if let value = options["blockPrivateNetworks"] as? Bool { blockPrivateNetworks = value }
        if let value = options["allowUriCredentials"] as? Bool { allowUriCredentials = value }
        if let value = options["respectServerCacheHeaders"] as? Bool {
            respectServerCacheHeaders = value
        }

        if options.keys.contains("allowedHosts") {
            allowedHosts = NextImageConfig.stringList(options["allowedHosts"])
        }
        if options.keys.contains("blockedHosts") {
            blockedHosts = NextImageConfig.stringList(options["blockedHosts"]) ?? []
        }

        if let value = NextImageConfig.positiveInt(options["maxUriLength"]) { maxUriLength = value }
        if let value = NextImageConfig.positiveInt(options["maxDataUriBytes"]) {
            maxDataUriBytes = value
        }
        if let value = NextImageConfig.positiveInt(options["maxHeaderCount"]) {
            maxHeaderCount = value
        }
        if let value = NextImageConfig.positiveInt(options["maxHeaderNameLength"]) {
            maxHeaderNameLength = value
        }
        if let value = NextImageConfig.positiveInt(options["maxHeaderValueLength"]) {
            maxHeaderValueLength = value
        }

        if let value = NextImageConfig.positiveDouble(options["memoryCacheBytes"]) {
            memoryCacheBytes = max(UInt(value), NextImageConfig.minMemoryCacheBytes)
        }
        if let value = NextImageConfig.positiveDouble(options["diskCacheBytes"]) {
            diskCacheBytes = max(UInt(value), NextImageConfig.minDiskCacheBytes)
        }
        if let value = NextImageConfig.positiveDouble(options["requestTimeoutMs"]) {
            requestTimeoutMs = min(max(value, 1000), 300_000)
        }

        if options.keys.contains("certificatePins") {
            certificatePins = NextImageConfig.parsePins(options["certificatePins"])
        }
    }

    private static func stringList(_ value: Any?) -> [String]? {
        guard let list = value as? [Any] else { return nil }
        return list.compactMap { $0 as? String }.filter { !$0.isEmpty }
    }

    private static func positiveInt(_ value: Any?) -> Int? {
        guard let number = value as? NSNumber else { return nil }
        let int = number.intValue
        return int > 0 ? int : nil
    }

    private static func positiveDouble(_ value: Any?) -> Double? {
        guard let number = value as? NSNumber else { return nil }
        let double = number.doubleValue
        return double > 0 ? double : nil
    }

    private static func parsePins(_ value: Any?) -> [String: [String]] {
        guard let map = value as? [String: Any] else { return [:] }
        var result: [String: [String]] = [:]
        for (host, pins) in map {
            guard let list = pins as? [Any] else { continue }
            let valid = list
                .compactMap { $0 as? String }
                .filter { NextImageSecurity.matches($0, pattern: pinPattern) }
            if !valid.isEmpty {
                result[host] = valid
            }
        }
        return result
    }
}

public struct NextImageParsedUri {
    public let scheme: String
    public let userInfo: String?
    public let host: String?
    public let port: Int?
    public let path: String
    public let query: String?
}

public enum NextImageUriResult {
    case allowed(uri: String, parsed: NextImageParsedUri)
    case blocked(code: String, message: String)
}

/// URL and header hardening, mirroring `src/security.ts` and
/// `NextImageSecurity.kt`. The JS layer validates first; this is the second
/// line of defence for anything reaching native directly.
public enum NextImageSecurity {
    private static let forbiddenHeaders: Set<String> = [
        "connection", "content-length", "expect", "host", "keep-alive",
        "proxy-authorization", "proxy-connection", "te", "trailer",
        "transfer-encoding", "upgrade", "via",
    ]

    private static let localSchemes: Set<String> = ["asset", "ph", "assets-library"]

    private static let uriPattern =
        "^([a-zA-Z][a-zA-Z0-9+.-]*):(//([^/?#]*))?([^?#]*)(\\?([^#]*))?(#(.*))?$"
    private static let headerNamePattern = "^[!#$%&'*+\\-.^_`|~0-9A-Za-z]+$"
    private static let ipv4Pattern = "^(\\d{1,3})\\.(\\d{1,3})\\.(\\d{1,3})\\.(\\d{1,3})$"

    static func matches(_ value: String, pattern: String) -> Bool {
        guard let regex = try? NSRegularExpression(pattern: pattern) else { return false }
        let range = NSRange(value.startIndex..., in: value)
        return regex.firstMatch(in: value, options: [], range: range) != nil
    }

    private static func captures(_ value: String, pattern: String) -> [String?]? {
        guard let regex = try? NSRegularExpression(pattern: pattern) else { return nil }
        let range = NSRange(value.startIndex..., in: value)
        guard let match = regex.firstMatch(in: value, options: [], range: range),
              match.range == range
        else {
            return nil
        }
        return (0 ..< match.numberOfRanges).map { index in
            let matchRange = match.range(at: index)
            guard matchRange.location != NSNotFound,
                  let swiftRange = Range(matchRange, in: value)
            else {
                return nil
            }
            return String(value[swiftRange])
        }
    }

    /// A strict URI parser. `URL` and `URLComponents` are lenient about hosts
    /// and credentials in ways that would let a blocked host through.
    public static func parseUri(_ uri: String) -> NextImageParsedUri? {
        guard let groups = captures(uri, pattern: uriPattern) else { return nil }

        let scheme = (groups[1] ?? "").lowercased()
        var userInfo: String?
        var host: String?
        var port: Int?

        if let authority = groups[3] {
            var rest = authority
            if let at = rest.lastIndex(of: "@") {
                userInfo = String(rest[rest.startIndex ..< at])
                rest = String(rest[rest.index(after: at)...])
            }

            if rest.hasPrefix("[") {
                guard let close = rest.firstIndex(of: "]") else { return nil }
                host = String(rest[rest.startIndex ... close]).lowercased()
                rest = String(rest[rest.index(after: close)...])
                if !rest.isEmpty, !rest.hasPrefix(":") { return nil }
            } else if let colon = rest.firstIndex(of: ":") {
                host = String(rest[rest.startIndex ..< colon]).lowercased()
                rest = String(rest[colon...])
            } else {
                host = rest.lowercased()
                rest = ""
            }

            if rest.hasPrefix(":") {
                let digits = String(rest.dropFirst())
                if !digits.isEmpty {
                    guard let parsedPort = Int(digits), parsedPort >= 1, parsedPort <= 65535 else {
                        return nil
                    }
                    port = parsedPort
                }
            }

            if host?.isEmpty == true { host = nil }
        }

        return NextImageParsedUri(
            scheme: scheme,
            userInfo: userInfo,
            host: host,
            port: port,
            path: groups[4] ?? "",
            query: groups[6]
        )
    }

    /// Supports `example.com`, `.example.com` and `*.example.com`.
    public static func hostMatches(_ host: String, pattern: String) -> Bool {
        let normalizedHost = host.lowercased().trimmingCharacters(in: CharacterSet(charactersIn: "[]"))
        var normalizedPattern = pattern.lowercased().trimmingCharacters(in: .whitespaces)

        if normalizedPattern.isEmpty { return false }
        if normalizedPattern == "*" { return true }
        if normalizedPattern.hasPrefix("*.") {
            normalizedPattern = String(normalizedPattern.dropFirst())
        }
        if normalizedPattern.hasPrefix(".") {
            let bare = String(normalizedPattern.dropFirst())
            return normalizedHost == bare || normalizedHost.hasSuffix(normalizedPattern)
        }
        return normalizedHost == normalizedPattern
    }

    /// Loopback, link-local, private-range, CGNAT and cloud metadata hosts.
    public static func isPrivateHost(_ host: String) -> Bool {
        var normalized = host.lowercased().trimmingCharacters(in: CharacterSet(charactersIn: "[]"))
        while normalized.hasSuffix(".") { normalized = String(normalized.dropLast()) }

        if normalized == "localhost"
            || normalized.hasSuffix(".localhost")
            || normalized.hasSuffix(".local")
            || normalized.hasSuffix(".internal")
        {
            return true
        }

        if let groups = captures(normalized, pattern: ipv4Pattern) {
            let octets = (1 ... 4).compactMap { groups[$0].flatMap(Int.init) }
            guard octets.count == 4 else { return true }
            if octets.contains(where: { $0 > 255 }) { return true }
            let a = octets[0]
            let b = octets[1]
            if a == 0 || a == 10 || a == 127 { return true }
            if a == 169, b == 254 { return true }
            if a == 172, (16 ... 31).contains(b) { return true }
            if a == 192, b == 168 { return true }
            if a == 100, (64 ... 127).contains(b) { return true }
            if a >= 224 { return true }
            return false
        }

        if normalized.contains(":") {
            if normalized == "::" || normalized == "::1" { return true }
            if normalized.hasPrefix("::ffff:") {
                return isPrivateHost(String(normalized.dropFirst("::ffff:".count)))
            }
            if matches(normalized, pattern: "^f[cd][0-9a-f]{0,2}:") { return true }
            if matches(normalized, pattern: "^fe[89ab][0-9a-f]?:") { return true }
        }

        return false
    }

    public static func validateUri(_ rawUri: String?, config: NextImageConfig) -> NextImageUriResult {
        guard let rawUri else {
            return .blocked(code: "EMPTY_URI", message: "Image source uri must not be empty.")
        }
        let uri = rawUri.trimmingCharacters(in: .whitespacesAndNewlines)
        if uri.isEmpty {
            return .blocked(code: "EMPTY_URI", message: "Image source uri must not be empty.")
        }
        if uri.count > config.maxUriLength {
            return .blocked(
                code: "URI_TOO_LONG",
                message: "Image source uri exceeds \(config.maxUriLength) characters."
            )
        }
        if uri.unicodeScalars.contains(where: { $0.value < 0x20 || $0.value == 0x7F }) {
            return .blocked(
                code: "CONTROL_CHARACTERS",
                message: "Image source uri contains control characters."
            )
        }

        guard let parsed = parseUri(uri) else {
            return .blocked(code: "MALFORMED_URI", message: "Image source uri is malformed.")
        }

        switch parsed.scheme {
        case "https":
            break
        case "http":
            if !config.allowInsecureHttp {
                return .blocked(
                    code: "INSECURE_SCHEME",
                    message: "Plain http:// sources are blocked by the NextImage security policy."
                )
            }
        case "data":
            if !config.allowDataUri {
                return .blocked(code: "DATA_URI_NOT_ALLOWED", message: "data: sources are blocked.")
            }
            if estimateDataUriBytes(uri) > config.maxDataUriBytes {
                return .blocked(
                    code: "DATA_URI_TOO_LARGE",
                    message: "data: source exceeds \(config.maxDataUriBytes) bytes."
                )
            }
            return .allowed(uri: uri, parsed: parsed)
        case "file":
            return config.allowFileUri
                ? .allowed(uri: uri, parsed: parsed)
                : .blocked(code: "FILE_URI_NOT_ALLOWED", message: "file: sources are blocked.")
        default:
            return localSchemes.contains(parsed.scheme)
                ? .allowed(uri: uri, parsed: parsed)
                : .blocked(
                    code: "SCHEME_NOT_ALLOWED",
                    message: "Unsupported uri scheme \"\(parsed.scheme)\"."
                )
        }

        if parsed.userInfo != nil, !config.allowUriCredentials {
            return .blocked(
                code: "URI_CREDENTIALS",
                message: "Image source uri must not embed credentials."
            )
        }
        guard let host = parsed.host else {
            return .blocked(code: "HOST_MISSING", message: "Image source uri has no host.")
        }
        if config.blockedHosts.contains(where: { hostMatches(host, pattern: $0) }) {
            return .blocked(code: "HOST_BLOCKED", message: "Host \"\(host)\" is blocked.")
        }
        if let allowed = config.allowedHosts,
           !allowed.contains(where: { hostMatches(host, pattern: $0) })
        {
            return .blocked(code: "HOST_NOT_ALLOWED", message: "Host \"\(host)\" is not allow-listed.")
        }
        if config.blockPrivateNetworks, isPrivateHost(host) {
            return .blocked(
                code: "PRIVATE_HOST_BLOCKED",
                message: "Host \"\(host)\" resolves to a private or loopback address."
            )
        }

        return .allowed(uri: uri, parsed: parsed)
    }

    /// Removes headers that could split the request or that URLSession owns.
    public static func sanitizeHeaders(
        _ headers: [(name: String, value: String)],
        config: NextImageConfig
    ) -> (headers: [String: String], order: [String], rejected: [String]) {
        var accepted: [String: String] = [:]
        var order: [String] = []
        var rejected: [String] = []

        for header in headers {
            if accepted.count >= config.maxHeaderCount {
                rejected.append(header.name)
                continue
            }
            if header.name.count > config.maxHeaderNameLength
                || !matches(header.name, pattern: headerNamePattern)
                || forbiddenHeaders.contains(header.name.lowercased())
            {
                rejected.append(header.name)
                continue
            }
            if header.value.count > config.maxHeaderValueLength
                || !isValidHeaderValue(header.value)
            {
                rejected.append(header.name)
                continue
            }
            if accepted.updateValue(header.value, forKey: header.name) == nil {
                order.append(header.name)
            }
        }

        return (accepted, order, rejected)
    }

    /// Decoded length of a data uri, without decoding it.
    private static func estimateDataUriBytes(_ uri: String) -> Int {
        guard let comma = uri.firstIndex(of: ",") else { return uri.count }
        let meta = String(uri[uri.startIndex ..< comma])
        let payload = String(uri[uri.index(after: comma)...])
        guard meta.lowercased().hasSuffix(";base64") else { return payload.count }
        let padding = payload.hasSuffix("==") ? 2 : (payload.hasSuffix("=") ? 1 : 0)
        return max(0, payload.count * 3 / 4 - padding)
    }

    private static func isValidHeaderValue(_ value: String) -> Bool {
        value.unicodeScalars.allSatisfy { scalar in
            scalar == "\t" || (scalar.value >= 0x20 && scalar.value <= 0x7E)
                || (scalar.value >= 0xA0 && scalar.value <= 0xFF)
        }
    }

    /// Credential-free, query-free rendering of a uri, safe for logs and events.
    public static func redactUri(_ uri: String) -> String {
        guard let parsed = parseUri(uri) else { return "<invalid-uri>" }
        if parsed.scheme == "data" { return "data:<redacted>" }
        let port = parsed.port.map { ":\($0)" } ?? ""
        let query = parsed.query == nil ? "" : "?<redacted>"
        return "\(parsed.scheme)://\(parsed.host ?? "")\(port)\(parsed.path)\(query)"
    }
}
